spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala

```diff
@@ -431,13 +432,99 @@ object DeltaHistoryManager extends DeltaLogging {
     override def getVersion: Long = version
   }
 
+  /**
+   * Abstract iterator for breaking a sequence of items into groups based on custom logic. It wraps
+   * an underlying iterator, transforming and grouping its items according to specific rules defined
+   * in subclasses. This iterator excels in scenarios requiring the ordered processing of data
+   * streams, where specific items mark the beginning of new groups. It ensures items within a group
+   * are emitted together.
+   */
+  private abstract class GroupBreakingIterator[T](
+      underlying: Iterator[T]) extends Iterator[ArrayBuffer[T]] {
+    private var bufferedOutput: Option[mutable.ArrayBuffer[T]] = None
+    private val bufferedUnderlying = underlying.buffered
+
+    protected def isNewGroupStart(item: T): Boolean
+
+    protected def transformItem(item: T): T
+
+    private def queueItemsIfNeeded(): Unit = {
+      if (bufferedOutput.isEmpty && bufferedUnderlying.hasNext) {
+        val group = new mutable.ArrayBuffer[T]()
+        while (bufferedUnderlying.hasNext &&
+          (!isNewGroupStart(bufferedUnderlying.head) || group.isEmpty)) {
+          group += transformItem(bufferedUnderlying.next())
+        }
+        bufferedOutput = Some(group)
+      }
+    }
+
+    override def hasNext: Boolean = {
+      queueItemsIfNeeded()
+      bufferedOutput.nonEmpty
+    }
+
+    override def next(): ArrayBuffer[T] = {
+      if (!hasNext) throw new NoSuchElementException("No more items to return")
+      val output = bufferedOutput.get
+      bufferedOutput = None
+      output
+    }
+  }
+
+  /**
+   * An iterator that helps select old log files for deletion. It takes the input iterator of
+   * eligible log file groups to delete, grouped by version and timestamp-adjusting dependency, and
+   * returns all groups up to but not including the last group containing a checkpoint.
+   * This ensures time-travel can be supported anywhere within the retention window.
+   * Example Scenario:
+   * - 25 commits made.
+   * - Checkpoint files at the 10th and 20th commits.
+   * - Retention window starts from the 15th commit.
+   * We keep everything from the 10th commit checkpoint onwards.
+   * Note 1: 10th commit is retained in the above scenario to get a valid timestamp for version 10.
+   * Note 2: This code assumes that underlying iterator list groups in a sorted-by-version order.
+   */
+  class LastCheckpointPreservingLogDeletionIterator(
+      underlying: Iterator[ArrayBuffer[ArrayBuffer[FileStatus]]])
+    extends Iterator[ArrayBuffer[ArrayBuffer[ArrayBuffer[FileStatus]]]]{
+
+    private class CheckpointGroupingIterator(
+        underlying: Iterator[ArrayBuffer[ArrayBuffer[FileStatus]]])
+      extends GroupBreakingIterator[ArrayBuffer[ArrayBuffer[FileStatus]]](underlying) {
```

Comment:
High level comment: These kinds of nested/buffered/grouped iterators make it very hard to see the overall flow of the algorithm -- there's not enough statement of intent. Some code comments explaining the overall flow at a high level, plus judicious use of type aliases would go a long way IMO, e.g.
```scala
  // All files with the same commit version
  type SameCommitVersionFileGroup = ArrayBuffer[FileStatus]
  
  // All commits whose timestamps need to be adjusted because of the first commit in the group
  type TimestampAdjustedCommitGroup = ArrayBuffer[SameCommitVersionFileGroup]
  
  // All timestamp groups that depend on the same checkpoint
  type DependentCheckpointGroup = ArrayBuffer[TimestampAdjustedCommitGroup]
  
  // Returns all checkpoint groups except the last, so we never delete the last checkpoint.
  class LastCheckpointPreservingLogDeletionIterator(
      underlying: Iterator[TimestampAdjustedCommitGroup])
    extends Iterator[DependentCheckpointGroup]{

    // Each next call returns all timestamp groups that depend on the same checkpoint.
    private class CheckpointGroupingIterator(
        underlying: Iterator[TimestampAdjustedCommitGroup])
      extends GroupBreakingIterator[TimestampAdjustedCommitGroup](underlying) {
```

Once such changes are made, it will be a lot easier to review the code for correctness by comparing against the stated intent.

Reply:
That def. makes sense.
1. Added the types
2. Added high level comments in the MetadataCleanup
3. Made the listExpiredDeltaLogs easier to read.